ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java

```diff
@@ -757,6 +767,73 @@ private AdminResponse handleCatchupStatusRequest(DataInputStream requestStream,
     return new CatchupStatusAdminResponse(isCaughtUp, adminResponse);
   }
 
+  /**
+   * Handles {@link com.github.ambry.protocol.AdminRequestOrResponseType#BlobStoreControl}.
+   * @param requestStream the serialized bytes of the request.
+   * @param adminRequest the {@link AdminRequest} received.
+   * @return the {@link AdminResponse} to the request.
+   * @throws IOException if there is any I/O error reading from the {@code requestStream}.
+   */
+  private AdminResponse handleBlobStoreControlRequest(DataInputStream requestStream, AdminRequest adminRequest)
+      throws IOException {
+    ServerErrorCode error;
+    BlobStoreControlAdminRequest blobStoreControlAdminRequest =
+        BlobStoreControlAdminRequest.readFrom(requestStream, adminRequest);
+    if (blobStoreControlAdminRequest.getNumReplicasCaughtUpPerPartition() > 0) {
+      PartitionId partitionId = blobStoreControlAdminRequest.getPartitionId();
+      if (partitionId != null) {
+        error = validateRequest(partitionId, RequestOrResponseType.AdminRequest);
+        if (error.equals(ServerErrorCode.No_Error)) {
+          if (blobStoreControlAdminRequest.shouldEnable()) {
+            // TODO: start BlobStore properly
+            error = ServerErrorCode.Temporarily_Disabled;
+          } else {
+            Collection<PartitionId> partitionIds = Collections.singletonList(partitionId);
+            if (storageManager.disableCompactionForBlobStore(partitionId)) {
+              controlRequestForPartitions(RequestOrResponseType.PutRequest, partitionIds, false);
```

Comment:
Does the failure detector honor `Temporary_Disabled`? Could you check and make that change as well?

Basically if the store is shut down, the failure detector should consider a replica as "unavailable". This is to avoid sending requests to these replicas. We can achieve similar effect with the change we proposed in Helix - to dynamically mark replica as unavailable, for which you created an issue - but the failure detector works orthogonal to that (it marks things based on responses to requests).

Take a look at the flow in `ResponseHandler.java` (who calls it and who it calls) to see how the failure detector works.

Reply:
+1 if this is not a whole lot of code. If not, I prefer that we add this issue to the helix one (since they are related) and address them separately from this change.

Reply:
In the current change, if the BlobStore is successfully stopped, then next request related to this BlobStore will return Disk_Unavailable error code because when validating the request, storageManager will first check if store is started. In this case, storageManager.getStore will return null which causes the Disk_Unavailable error. Hence, we are fine in current stage.